gravitee-am-management-api/gravitee-am-management-api-service/src/main/java/io/gravitee/am/management/service/impl/upgrades/ApplicationFactorSettingsUpgrader.java

```diff
@@ -0,0 +1,191 @@
+                    if (!isEmpty(app.getFactors())) {
+                        return factorService.findByDomain(app.getDomain())
+                                .filter(factor -> factor.getFactorType().equals(FactorType.RECOVERY_CODE))
+                                .map(Factor::getId)
+                                .toList()
+                                .flatMap(recoveryCodeIds -> {
+                                    // transfer the list of factors into the new Factor Settings structure
+                                    // consider the first factor of the list as the default one.
+                                    moveFactorIdsIntoFactorSettings(app, recoveryCodeIds);
+
+                                    if (app.getSettings() != null && app.getSettings().getMfa() != null) {
+                                        migrateEnrollSettings(app);
+                                        migrateChallengeSettings(app);
+                                        migrateStepUpSettings(app);
+                                    }
+
+                                    logger.debug("Update factor settings for application '{}'", app.getId());
+                                    return applicationService.update(app);
+                                });
+                    } else {
+                        logger.debug("No factor to process for application '{}'", app.getId());
+                    }
+                    return Single.just(app);
+                }).ignoreElements()
+                .doOnError(err -> updateSystemTask(task, (SystemTaskStatus.FAILURE), task.getOperationId()).subscribe())
+                .andThen(updateSystemTask(task, SystemTaskStatus.SUCCESS, task.getOperationId())
+                        .map(__ -> true)
+                        .onErrorResumeNext((err) -> {
+                            logger.error("Unable to update status for task {}: {}", TASK_ID, err.getMessage());
+                            return Single.just(false);
+                        }))
+                .onErrorResumeNext((err) -> {
+                    logger.error("Unable to migrate factor settings for applications: {}", err.getMessage());
+                    return Single.just(false);
+                });
+    }
+
+    private static MFASettings migrateEnrollSettings(Application app) {
+        final var mfaSettings = app.getSettings().getMfa();
+        final var enrollSettings = new EnrollSettings();
+        mfaSettings.setEnroll(enrollSettings);
+        enrollSettings.setActive(true);
+        if (!mfaSettings.getEnrollment().getForceEnrollment()) {
+            enrollSettings.setType(MfaEnrollType.OPTIONAL);
+            enrollSettings.setForceEnrollment(false);
+            enrollSettings.setSkipTimeSeconds(mfaSettings.getEnrollment().getSkipTimeSeconds());
+        } else {
+            enrollSettings.setType(MfaEnrollType.REQUIRED);
+            enrollSettings.setForceEnrollment(true);
+        }
```

Comment:
MfaEnrollType.CONDITIONAL is missing

Reply:
In 4.2 and previous version, enrollment is either REQUIRED or OPTIONAL. When CONDITIONAL rule is defined then it apply only on challenge.